File: leetcode/Minimum Path Sum.py
class Solution:
    # @param grid, a list of lists of integers
    # @return an integer
    def minPathSum(self, grid):
        m, n = len(grid)-1, len(grid[0])-1
        table = [[]] * (m+1)
        for i in range(m+1):
            table[i] = [None] * (n+1)
            
        def fun(mm, nn):
            if mm == m and nn == n:
                return grid[mm][nn]
            if table[mm][nn]:
                return table[mm][nn]
            a, b = 4294967296, 4294967296
            #down
            if mm + 1 <= m:
                a = fun(mm+1, nn)
            #right
            if nn + 1 <= n:
                b = fun(mm, nn+1)
            table[mm][nn] = min(a, b) + grid[mm][nn]
            return table[mm][nn]
        if grid:
            return fun(0,0)
        else:
            return 0

File: leetcode/cpp/Minimum_Path_Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if (grid.size() == 0)
            return 0;
        int row = grid.size();
        int col = grid[0].size();
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (0 == i && 0 == j)
                    continue;
                if (i - 1 < 0)
                    grid[i][j] += grid[i][j - 1];
                else if (j - 1 < 0)
                    grid[i][j] += grid[i - 1][j];
                else
                    grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]);
            }
        }
        
        return grid[row - 1][col - 1];
    }
};